### PR TITLE
[Frontend] Support CustomToolCall in the non-Harmony Responses API

### PR DESCRIPTION
**Pull request by Dovis01**
## Purpose

First of the smaller PRs splitting #56285 per review ("I'd suggest prioritizing support for CustomToolCall first"). @chaunceyjiang

`type=custom` tools are presented to the shared function-tool parsers as a single-string `input` function shim: the schema is generated by `custom_tool_parameters`, a grammar `format` is folded into the description, and schema lookup / xgrammar structural tags / named `tool_choice` with `type=custom` all see the same shim (`vllm/tool_parsers/utils.py`, `structural_tag_registry.py`, `vllm/parser/abstract_parser.py`, and the `tool_choice` validator in `responses/protocol.py`). Completed calls stream as `response.custom_tool_call_input.delta` / `.done` events (with a streaming-safe partial decoder, including surrogate-pair handling, so the concatenated deltas are always a prefix of the final input) and materialize as `ResponseCustomToolCall` items with `ctc_` ids; replaying a `custom_tool_call` / `custom_tool_call_output` item re-encodes the payload through the same shim, so multi-turn Codex-style sessions round-trip with no server-side storage.

**Overlap disclosure.** #56418 ("Preserve custom_tool_call on non-Harmony Responses routes", open since 2026-09-11) implements the same shim approach and its own description already notes the overlap with #56285. This PR is the CustomToolCall subset of #56285 and additionally covers the `ToolChoiceCustom` acceptance paths in `DelegatingParser` (forced/named tool choice, required-tool streaming extraction) and tests for forced custom tools against the GLM 4.7 structural-tag parser; #56418 additionally covers `tool_choice=none` and namespace flattening. Whichever lands first, the other can be reduced to the remainder — happy to rebase or fold whichever direction maintainers prefer. #49882 proposes the opposite (reject custom tools) and is superseded by either implementation.

## Tests

```
pytest tests/entrypoints/openai/responses/test_responses_utils.py \
       tests/entrypoints/openai/responses/test_streaming_events.py \
       tests/entrypoints/openai/responses/test_serving_responses.py \
       tests/tool_parsers/test_structural_tag_registry.py \
       tests/tool_use/test_tool_choice_required.py -q
# 303 passed, 7 skipped (+ 49 passed / 1 xfailed on the serving file)
```

## Real-model validation

GLM-5.3-Flash on 8xH100 served from this branch (same flags as the Codex validation below):

- The four malformed `tools` bodies from the review each return **HTTP 400** with a pydantic field-validation error (`list_type` / `model_attributes_type` / `literal_error`), not a 500; server log clean.
- A grammar-format custom tool returns **HTTP 400** with "Custom tools with a grammar format are not supported; only free-form text custom tools are accepted."
- Named `tool_choice={"type":"custom","name":"emit_command"}` streaming: `custom_tool_call` item with `ctc_` id, concatenated `custom_tool_call_input.delta`s equal the `.done` input (`pwd`), added/done ids match, payload non-empty.
- `custom_tool_call` + `custom_tool_call_output` replay returns 200 and the follow-up answer carries the tool result verbatim.

## Real Codex validation

Real Codex CLI 0.154.0 sessions (`codex exec --profile glm53`, provider `wire_api="responses"`, `store=false`, `stream=true`, `include=["reasoning.encrypted_content"]`, 12 tools including `namespace` and `web_search`, ~17 KB `instructions`, `parallel_tool_calls=true`), all traffic captured by a local logging proxy in front of the server:
  * Debugged a broken calculator: rewrote `calc.py` as a recursive-descent parser with precedence, parentheses and unary minus; 7/7 tests pass; 15 requests.
  * Built a terminal snake game from scratch: `snake.py` (curses UI over a curses-free logic class), `test_snake.py` with 44 passing tests, `README.md`, `smoke_test.py` (pty launch, keys, `q`, clean exit; exit code 0). 27 requests, history up to 106 input items / 189 KB per request.
  * Proxy log: 48/48 Codex requests HTTP 200 (42 `POST /v1/responses`, 6 `GET /v1/models`), zero 4xx/5xx; every replayed turn carried `reasoning` items with `encrypted_content`, `function_call` and `function_call_output` items; `output_item.done` items matched the  final `output` on the sampled streams. vLLM server log: 0 tracebacks. One stream was  closed by the client mid-reasoning (Codex restarted its thread); the server answered 200 and logged nothing.

## Known limitation

Custom tools do not honor `--tool-strict-level`: `_dump_tool_for_xgrammar` emits no `strict` for a `CustomTool` and `_with_tool_strict` only pins `FunctionTool` / `ChatCompletionToolsParam`, so under `strict_level=parameter` a custom tool's `{"input": ...}` arguments stay unconstrained while sibling function tools are pinned. Nothing that worked before breaks; full strict support can be a follow-up.

New: custom tool output item + history replay through the shim + input decode/partial-decode (incl. surrogate pairs); `custom_tool_call_input` streaming events end-to-end through the processor; `arguments.done` emitted even without prior deltas; forced `tool_choice` (required and named `type=custom`) sees custom tools in JSON schema and structural tags, incl. the GLM 4.7 parser. Unchanged from main: encrypted-content rejection behavior, function-call streaming, serving paths.

### CODE REVIEW

File: vllm/entrypoints/openai/responses/utils.py [hunk 1]
```diff
@@ -50,6 +53,109 @@
...
+    "/": "/",
+    "b": "\b",
+    "f": "\f",
+    "n": "\n",
+    "r": "\r",
+    "t": "\t",
+}
+
+
+def _decode_unicode_escape(buffer: str, i: int) -> tuple[str, int] | None:
```
Comment (chaunceyjiang) [inline vllm/entrypoints/openai/responses/utils.py]: Are there any other approaches? Manually writing the decode logic like this may be difficult to understand.

File: vllm/tool_parsers/utils.py [hunk 2]
```diff
@@ -35,6 +38,44 @@
 
 logger = init_logger(__name__)
 
+CUSTOM_TOOL_INPUT_KEY = "input"
```
Comment (chaunceyjiang) [inline vllm/tool_parsers/utils.py:40]: I think the bigger concern here is that we're effectively modeling a custom tool as a function tool with an `{"input": string}` schema.

https://developers.openai.com/api/docs/guides/function-calling?api-mode=responses

One of the main semantics of a custom tool is that its input is free-form text and does not need to be wrapped in JSON. More importantly, `format=grammar` (CFG)  is supposed to constrain the generated custom-tool input, while this implementation only adds the grammar to the tool description, so it cannot actually enforce that constraint.

The handwritten partial JSON decoder also seems to be a consequence of this shim: we first force the free-form payload into JSON, then need to decode it again while streaming.

Would it make sense to either support custom-tool input natively in the parser/structured-output path, or initially support only plain custom tools and explicitly leave `format=grammar` unsupported?

Comment (Dovis01) [reply vllm/tool_parsers/utils.py:40]: Thanks, both points taken. 

In dda4753797: 
(1) format=grammar custom tools are now rejected at request validation with a clear 400 instead of folding the grammar into the description.
(2) The hand-written partial decoder is gone; the streamed input prefix is decoded with partial_json_parser (already used by the tool parsers), with a one-line guard that drops a trailing lone high surrogate so streamed deltas stay a prefix of the final input. I kept the {"input": string} shim itself for this PR: making the parsers and chat templates emit free-form custom-tool input natively touches every tool parser, so I'd propose that as a follow-up once the plain-text path is in.

**Review by chaunceyjiang — COMMENTED** [review]
Thanks, this looks much clearer now. Explicitly rejecting format=grammar and reusing partial_json_parser address my main concerns with the previous implementation.

File: vllm/entrypoints/openai/responses/protocol.py [hunk 3]
```diff
@@ -602,6 +604,25 @@ def input_item_parsing(cls, data):
         data["input"] = processed_input
         return data
 
+    @model_validator(mode="before")
+    @classmethod
+    def check_custom_tool_format(cls, data):
+        if not isinstance(data, dict):
+            return data
+        for tool in data.get("tools") or []:
+            tool_dict = tool if isinstance(tool, dict) else tool.model_dump()
```
Comment (sfeng33) [inline vllm/entrypoints/openai/responses/protocol.py]: This guards `data` being a dict, but not the shape of `tools` or of each entry, so a malformed body now raises a bare `AttributeError` (HTTP 500) instead of the pydantic `ValidationError` (400/422) it produced before this validator existed. Each of these returns a 4xx on `main` and a 500 with the patch:

- `tools="abc"` — iterating the string yields `"a"`, and `"a".model_dump()` raises `AttributeError: 'str' object has no attribute 'model_dump'`.
- `tools={"type": "custom"}` — iterating the dict yields the key `"type"`, same failure.
- `tools=[None]` — `None.model_dump()`.
- `tools=[{"type": "custom", "name": "x", "format": "text"}]` — `format` is a string, so `tool_format = "text" or {}` is `"text"` and the next line hits `AttributeError: 'str' object has no attribute 'get'`.

Because this runs `mode="before"`, it fires ahead of the field validation that would otherwise reject these cleanly, so any client with a slightly wrong body gets a 500 rather than a message telling them what's wrong.

File: vllm/parser/abstract_parser.py [hunk 4]
```diff
@@ -300,7 +302,7 @@ def _extract_tool_calls(
         supports_required_and_named = tool_parser.supports_required_and_named
         is_named_tool_choice = request.tool_choice and isinstance(
             request.tool_choice,
-            (ToolChoiceFunction, ChatCompletionNamedToolChoiceParam),
+            (ToolChoiceFunction, ToolChoiceCustom, ChatCompletionNamedToolChoiceParam),
```
Comment (sfeng33) [inline vllm/parser/abstract_parser.py:305]: Accepting `ToolChoiceCustom` here (and in the `ResponsesRequest` validator) makes a request shape reachable that the validator previously rejected, but the per-model parsers that special-case named tool choice weren't updated alongside it:

- `gemma4_engine_tool_parser.py:31`
- `poolside_v1_tool_parser.py:174`
- `muse_glimmer_tool_parser.py:231`
- `kimi_k3_tool_parser.py:152`
- `k2_horizon_tool_parser.py:187`

Each still checks `isinstance(tc, (ChatCompletionNamedToolChoiceParam, ToolChoiceFunction))`, so a `ToolChoiceCustom` falls off the intended path. Traced for gemma4: it has no `structural_tag_model` (`get_model_structural_tag(model="gemma4", ...)` raises `Unknown format type: gemma4`), so `_apply_structural_tag` returns early; `Gemma4EngineToolParser.adjust_request` then fails its isinstance check, skips `skip_special_tokens = False`, and falls through to the base `adjust_request`, which forces JSON guided decoding — the path that method's own docstring describes as leaking as content and crashing EngineCore under speculative decoding.

Either widen those five checks in this PR, or keep rejecting `ToolChoiceCustom` for parsers that can't route it yet. The latter is smaller if you'd rather not touch five parsers here.

File: vllm/entrypoints/openai/responses/streaming_events.py [hunk 5]
```diff
@@ -1068,9 +1111,31 @@ def emit_simple_tool_call_delta(
 def emit_simple_tool_call_done(
     state: SimpleStreamingState,
 ) -> list[StreamingResponsesResponse]:
-    events: list[StreamingResponsesResponse] = []
-    if state.has_emitted_tool_call_delta:
+    events: list[StreamingResponsesResponse]
+    item: ResponseOutputItem
+    if state.tool_call_is_custom:
+        payload = decode_custom_tool_input(state.accumulated_text)
+        if not payload.startswith(state.tool_call_payload):
```
Comment (sfeng33) [inline vllm/entrypoints/openai/responses/streaming_events.py]: This fallback reports a truncated fragment as the complete input rather than surfacing that something went wrong. When the final decode isn't an extension of what was already streamed, `payload` becomes `state.tool_call_payload` — whatever prefix happened to be emitted — and that value goes out in both `custom_tool_call_input.done` and the `output_item.done` item.

Reproduced through `SimpleStreamingEventProcessor` with two chunks, `'{"a": "xy"'` then `', "input": "hello"}'`:

- streaming emits `custom_tool_call_input.done(input="xy")` and an output item with `input="xy"`
- `decode_custom_tool_input` on the same complete arguments returns `"hello"`

So the client is handed `"xy"` — a fragment of an unrelated key — marked complete, and the streaming and non-streaming paths disagree on identical model output. `_emit_custom_tool_input_delta` compounds it by dropping any non-prefix delta, so nothing upstream ever signals the divergence.

The mismatch is only reachable because of the single-key fallback in `_custom_tool_payload` (see my comment on `utils.py`); fixing that one removes most of the divergence. For whatever remains, I'd rather see the authoritative final `decode_custom_tool_input` value win — emit the correcting delta even when it isn't a clean extension — than have `done` report a prefix it knows is wrong.

File: vllm/entrypoints/openai/responses/utils.py [hunk 6]
```diff
@@ -50,6 +55,53 @@
...
+
+def encode_custom_tool_input(payload: str) -> str:
+    return json.dumps({CUSTOM_TOOL_INPUT_KEY: payload}, ensure_ascii=False)
+
+
+def _custom_tool_payload(parsed: Any) -> str | None:
+    if not isinstance(parsed, dict):
+        return None
+    value = parsed.get(CUSTOM_TOOL_INPUT_KEY)
+    if not isinstance(value, str) and len(parsed) == 1:
```
Comment (sfeng33) [inline vllm/entrypoints/openai/responses/utils.py]: This fallback unwraps *any* single-key object, not just the `input` shim, which defeats the bare-text pass-through the docstring on `decode_custom_tool_input` promises just below. A custom tool exists precisely to carry a freeform payload, so a model that emits its payload directly rather than through the `{"input": ...}` shim gets silently rewritten whenever that payload is a one-key JSON object:

```python
>>> decode_custom_tool_input('{"file": "a.py"}')
'a.py'
```

The tool receives `a.py` instead of the JSON object the model actually produced, and there's no way for the caller to tell the two apart afterwards. Unwrapping only the exact `CUSTOM_TOOL_INPUT_KEY` and passing everything else through keeps the shim working and preserves the documented behavior:

```python
def _custom_tool_payload(parsed: Any) -> str | None:
    if not isinstance(parsed, dict):
        return None
    value = parsed.get(CUSTOM_TOOL_INPUT_KEY)
    return value if isinstance(value, str) else None
```

This also removes the main way streaming and non-streaming can disagree (see my comment on `streaming_events.py:1118`). If the fallback is deliberate — covering a model that wraps the payload under its own key — it'd help to say which model in a comment, since as written it fires for every single-key payload.

File: vllm/tool_parsers/structural_tag_registry.py [hunk 7]
```diff
@@ -209,6 +210,11 @@ def _dump_tool_for_xgrammar(
         if tool.strict is not None:
             function["strict"] = tool.strict
         return {"type": "function", "function": function}
+    if isinstance(tool, CustomTool):
```
Comment (sfeng33) [inline vllm/tool_parsers/structural_tag_registry.py:213]: Non-blocking, but worth recording: this branch never emits `"strict"`, and `_with_tool_strict` (line 115) still handles only `FunctionTool` / `ChatCompletionToolsParam`, so `_resolve_tool_strictness` can't pin a custom tool.

With `strict_level=ToolStrictLevel.PARAMETER` and `tool_choice="required"`, a function tool dumps as `{"name": "f", ..., "strict": true}` while a custom tool in the same request dumps with no `strict` key at all. Under `--tool-strict-level parameter` the custom tool's `{"input": ...}` arguments stay unconstrained while every sibling function tool is pinned — which is also what lets arbitrary text reach the decode paths I flagged in `utils.py` and `streaming_events.py`.

Nothing that worked before breaks, so this seems fine as a follow-up; I'd just call it out in the PR description so custom tools aren't assumed to honor the strict level.

**Comment by sfeng33** [comment]
Please test the PR with a real model, unit test is not enough in this case.

**Comment by Dovis01** [comment]
> Please test the PR with a real model, unit test is not enough in this case.

Hi, @sfeng33. Real model verified.

Thanks a lot for the thorough review and for tracing these edge cases in detail. I've addressed all five comments and pushed the updates.

The changes include tightening the mode="before" validation path, handling ToolChoiceCustom consistently across the affected parsers, removing the overly broad single-key custom-tool fallback, making the final decoded custom-tool input authoritative in streaming, and aligning/documenting the custom-tool strictness behavior.

Really appreciate the detailed reproductions and suggestions — they were very helpful in cleaning up the implementation. I’ll also consider more these kinds of edge cases and boundary conditions thoroughly in future changes.
